sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1654,6 +1655,43 @@ class Dataset[T] private[sql](
     new MergeIntoWriterImpl[T](table, this, condition)
   }
 
+  /**
+   * Update rows in a table.
+   *
+   * Scala Example:
+   * {{{
+   *   spark.table("source")
+   *    .update(Map("salary" -> lit(200)))
+   *    .execute()
+   * }}}
+   * @param assignments A Map of column names to Column expressions representing the updates
+   *     to be applied.
+   * @group basic
+   * @since 4.0.0
+   */
+  def update(assignments: Map[String, Column]): Unit = {
+    updateInternal(assignments)
+  }
+
+  /**
+   * Update rows in a table that match a condition.
+   *
+   * Scala Example:
+   * {{{
+   *   spark.table("source")
+   *    .update(Map("salary" -> lit(200)), $"salary" === 100)
+   *    .execute()
+   * }}}
+   * @param assignments A Map of column names to Column expressions representing the updates
+   *     to be applied.
+   * @param condition the update condition
+   * @group basic
+   * @since 4.0.0
+   */
+  def update(assignments: Map[String, Column], condition: Column): Unit = {
+    updateInternal(assignments, Some(condition))
+  }
```

Comment:
Why is this needed when the condition could already be applied on the input that feeds into the DataFrame writer?

Reply:
`df.filter(...).update(...)` looks a bit weird as `filter` should be used to filter the input rows instead of the UPDATE condition. It also complicates the implementation as we need to extract filters to build the `UpdateTable` command.

Reply:
Semantic-wise, UPDATE TABLE reads some rows from the table w.r.t. the condition, performs updates, and writes back the updated rows to the table. `df.filter(...).update(...)` doesn't look intuitive for the "write back" part.

Reply:
BTW, this is also how Delta Lake defines the scala update API: https://github.com/delta-io/delta/blob/master/spark/src/main/scala/io/delta/tables/DeltaTable.scala#L280

Reply:
Yes in my thought, update needs to write back to the table `df`. but here ```df.filter() ``` (the object update() is called on) is a temporary dataset rather than a persisted table.  Ie, in SQL ``` UPDATE (select * from table where ... ) SET... ``` doesn't work

Reply:
I think we get side-tracked maybe a bit too much by mapping this into "one" particular SQL query instead of looking at this from a usability perspective.

IMHO, there are couple of ways to look at this:

```
spark.table("simple").write.update(mapping, cond)
```

This one is in line with the API of the existing datagram writer.

```
spark.table("simple").update()
```

This one doesn't follow the builder pattern of the writer and is much more a new API.

```
spark.catalog.getTable().update()
```

This one seems to be a bit lazy because it tries to trade the complexity of dealing with the analysis error handling of the input relation with a weirder API.

Why not just keep it in the actual dataframe writer API?

Reply:
`spark.table("simple").write.update(mapping, cond)` This looks really weird to me. Normally the DataFrame before `.write` is the input data to write out, but now it becomes the target to write data into. I believe this is counterintuitive to many Spark users.

What's your concern for `spark.catalog.getTable(...).update(...)`? It basically uses a dedicated `Table` API instead of DataFrame to host the update API. And we already have other DDL APIs there, such as `spark.catalog.createTable`, `spark.catalog.listTables`, etc.

Reply:
Im ok with exploring spark.catalog.getTable().update() for its compile time safety, if there is some agreement?

spark.table.update() is simpler than that, but it would probably make more sense if this returned a dataset for subsequent operation instead of replacing the dataset (as the implementation today is the latter)?

Reply:
@cloud-fan i took a look at this approach, the Table class doesnt have Catalog object and I have to pass it in , is it the right way?  (Sorry not too familiar here)

Reply:
Yea we can put some internal fields to this `Table` class that are necessary to construct the UPDATE command.

Reply:
Is this the right place to have the discussion on where to put the API? Is it worth it to have a very small doc outlining the alternatives and the intent and having the discussion on the mailing list instead?

Reply:
SGTM, let's have a broader discussion on dev list

Reply:
OK let me prepare this.